evwatcher.c

```diff
@@ -0,0 +1,76 @@
+/*
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. The name of the author may not be used to endorse or promote products
+ *    derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "event2/event_watcher.h"
+#include "event-internal.h"
+
+static inline struct event_watcher *
+event_watcher_new(struct event_base *base, union event_watcher_cb callback, unsigned type)
+{
+	struct event_watcher *watcher = mm_malloc(sizeof(struct event_watcher));
+	if (!watcher)
+		return NULL;
+	watcher->base = base;
+	watcher->type = type;
+	watcher->callback = callback;
+	TAILQ_INSERT_TAIL(&base->watchers[type], watcher, next);
+	return watcher;
+}
+
+struct event_watcher *
+event_watcher_prepare_new(struct event_base *base, event_prepare_cb callback)
+{
+	union event_watcher_cb cb = { .prepare = callback };
+	return event_watcher_new(base, cb, EV_WATCHER_PREPARE);
+}
+
+struct event_watcher *
+event_watcher_check_new(struct event_base *base, event_check_cb callback)
+{
+	union event_watcher_cb cb = { .check = callback };
+	return event_watcher_new(base, cb, EV_WATCHER_CHECK);
+}
+
+struct event_base *
+event_watcher_base(struct event_watcher *watcher)
+{
+	return watcher->base;
+}
+
+void
+event_watcher_free(struct event_watcher *watcher)
+{
+	TAILQ_REMOVE(&watcher->base->watchers[watcher->type], watcher, next);
+	mm_free(watcher);
+}
+
+int
+event_prepare_get_timeout(const struct event_prepare_cb_info *info, struct timeval *timeout)
```

Comment:
hm, looking into this again, and `event_prepare_get_timeout`does not looks like a good naming, how about `evwatch_prepare_get_timeout()` ? (and convert the whole patch into this)

Reply:
Do you mean to change all:
```C
struct event_watcher => struct evwatch
struct event_prepare_cb_info => struct evwatch_prepare_cb_info
struct event_check_cb_info => struct evwatch_check_cb_info

event_prepare_cb => evwatch_prepare_cb
event_check_cb => evwatch_check_cb

event_watcher_prepare_new => evwatch_prepare_new
event_watcher_check_new => evwatch_check_new
event_watcher_base => evwatch_base
event_watcher_free => evwatch_free
```
Should I also rename `event_watcher.h` to `evwatch.h`? I don't understand the naming conventions here... Or do you just want me to rename the one function?

Reply:
Yep, all of them, i.e. like `evbuffer` was done
And if we are doing this like `evbuffer`, then the file should be called `<event2/watch.h>`